src/main/native/org_cryptomator_windows_autostart_WinShortcutCreation_Native.cpp

```diff
@@ -0,0 +1,117 @@
+#ifndef UNICODE
+    #define UNICODE
+#endif
+
+#include <jni.h>
+#include <windows.h>
+#include <winnls.h>
+#include <shobjidl.h>
+#include <objbase.h>
+#include <objidl.h>
+#include <shlguid.h>
+#include <wchar.h>
+#include "org_cryptomator_windows_autostart_WinShortcutCreation_Native.h"
+
+LPCWSTR get_nullterminate_and_release(JNIEnv * env,  jstring name);
+HRESULT CreateLink(LPCWSTR lpszPathObj, LPCWSTR lpszPathLink, LPCWSTR lpszDesc);
+
+JNIEXPORT jint JNICALL Java_org_cryptomator_windows_autostart_WinShortcutCreation_00024Native_createShortcut
+  (JNIEnv * env, jobject thisObj, jstring target, jstring storage_path, jstring description) {
+
+    //make the utf16 char arrays null terminated
+    LPCWSTR link_target = get_nullterminate_and_release(env,target);
+    LPCWSTR link_location = get_nullterminate_and_release(env,storage_path);
+    LPCWSTR link_description = get_nullterminate_and_release(env,description);
```

Comment:
LPCWSTR is just `typedef const wchar_t*`, which [MS tells us the following about](https://docs.microsoft.com/en-us/cpp/cpp/char-wchar-t-char16-t-char32-t):

> The wchar_t type is an implementation-defined wide character type. In the Microsoft compiler, it represents a 16-bit wide character used to store Unicode encoded as UTF-16LE, the native character type on Windows operating systems.

Therefore we should explicitly encode our strings with UTF-16-LE, which is probably easiest to do in Java: Change the method signature to `JNIEnv * env, jobject thisObj, jbyteArray targetBytes, jbyteArray storagePathBytes, jbyteArray descriptionBytes)` and pass in `str.getBytes(StandardCharsets.UTF_16LE)`.

Now you _should_ be able to just cast the (single-width) byte array our target (double width) char array:

`LPCWSTR target = (wchar_t*) env->GetByteArrayElements(targetBytes, NULL);`

Reply:
Oh and add a null terminator, for primitive strings, of course.

Reply:
Resolved in 9cd216b